annotation-processor/src/main/java/org/hibernate/validator/ap/ConstraintAnnotationVisitor.java

```diff
@@ -191,16 +195,19 @@ public Void visitTypeAsInterface(TypeElement e, List<AnnotationMirror> p) {
 	 * @param mirrors The annotations to check.
 	 */
 	private void checkConstraints(Element annotatedElement, List<AnnotationMirror> mirrors) {
-
 		for ( AnnotationMirror oneAnnotationMirror : mirrors ) {
 
 			try {
 
 				ConstraintChecks constraintChecks = constraintCheckFactory.getConstraintChecks(
 						annotatedElement, oneAnnotationMirror
 				);
-				Set<ConstraintCheckError> errors = constraintChecks.execute( annotatedElement, oneAnnotationMirror );
-				messager.reportErrors( errors );
+				Set<ConstraintCheckIssue> errors = constraintChecks.execute( annotatedElement, oneAnnotationMirror );
+
+				Map<Class, Set<ConstraintCheckIssue>> issues = errors.stream().collect( Collectors.groupingBy( issue -> issue.getClass(), Collectors.toSet() ) );
```

Comment:
Instead of the two sub-types, couldn't you just add something like `getKind()` to `ConstraintCheckError` (renaming it is still a good idea though) an then just use this kind when reporting the diagnostic?